src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupLoggingHelper.cs

```diff
@@ -281,6 +287,60 @@ private static void AppendStringFromParameterValue(ReuseableStringBuilder sb, ob
             return args;
         }
 
+        private static void CreateItemsSnapshot(ref IList items)
+        {
+            if (items == null)
+            {
+                return;
+            }
+
+#if FEATURE_APPDOMAIN
+            // If we're in the default AppDomain, but any of the items come from a different AppDomain
+            // we need to take a snapshot of the items right now otherwise that AppDomain might get
+            // unloaded by the time we want to consume the items.
+            // If we're not in the default AppDomain, always take the items snapshot.
+            //
+            // It is unfortunate to need to be doing this check, but ResolveComReference and other tasks
+            // still use AppDomains and create a TaskParameterEventArgs in the default AppDomain, but
+            // pass it Items from another AppDomain.
+            if (AppDomain.CurrentDomain.IsDefaultAppDomain())
+            {
+                bool needsSnapshot = false;
+                foreach (var item in items)
+                {
+                    if (RemotingServices.IsTransparentProxy(item))
+                    {
+                        needsSnapshot = true;
+                        break;
+                    }
+                }
+
+                if (!needsSnapshot)
+                {
+                    return;
+                }
+            }
+#endif
+
+            int count = items.Count;
```

Comment:
Do we need this if not FEATURE_APPDOMAIN?

Reply:
oh, great point. I think if APPDOMAIN is not defined we should just exit the entire method

Reply:
If APPDOMAIN is not defined, we won't be in this method anyway since the callsite is also conditional.

Reply:
That additional #if is just to make the compiler happy. Technically we could make the entire method conditional.